src/main/java/uk/gov/hmcts/reform/sscs/service/exceptions/FeignClientErrorDecoder.java

```
@@ -0,0 +1,89 @@
+        this.appInsightsService = appInsightsService;
+    }
+
+    @Override
+    public Exception decode(String methodKey, Response response) {
+        Request originalRequest = response.request();
+        HttpMethod httpMethod = originalRequest.httpMethod();
+        HmcFailureMessage failMsg = null;
+
+        if (httpMethod.equals(HttpMethod.POST) || httpMethod.equals(HttpMethod.PUT)) {
+            HearingRequestPayload payload = mapToPostHearingRequest(originalRequest);
+            if (payload != null) {
+                failMsg = HmcFailureMessage.builder()
+                    .requestType(httpMethod.toString())
+                    .caseID(Long.valueOf(payload.getCaseDetails().getCaseRef()))
+                    .timeStamp(LocalDateTime.now())
+                    .errorCode(String.valueOf(response.status()))
+                    .errorMessage(response.reason())
+                    .build();
+            }
+        } else {
+            Long caseId = Long.parseLong(response.request().requestTemplate().queries().get("id").iterator().next());
+            failMsg = HmcFailureMessage.builder()
+                .requestType(httpMethod.toString())
+                .caseID(caseId)
+                .timeStamp(LocalDateTime.now())
+                .errorCode(String.valueOf(response.status()))
+                .errorMessage(response.reason())
+                .build();
+        }
+
+        switch (response.status()) {
+            case 400:
+            case 401:
+            case 403:
+            case 404: {
+                log.error("Error in calling Feign client. Status code "
+                              + response.status() + ", methodKey = " + methodKey);
+                log.error("Error details: {}", response.body().toString());
+                try {
+                    appInsightsService.sendAppInsightsEvent(failMsg);
+                } catch (JsonProcessingException e) {
+                    log.error("Error sending app insight event");
```

I think we should attempt to write a generic error message containing the case ID and bare minimum information. If that fails then we should throw a runtime exception.